### PR TITLE
DEVPROD-7656: Create Packages table on Build Information page

### PR DESCRIPTION
**Pull request by sophiabarness**
DEVPROD-7656

### Description

Created the Packages table on the Build Information page. It supports pagination and filtering.

### Screenshots

Packages Table

### Testing

Added Jest tests to ensure table is properly rendered.

### CODE REVIEW

File: apps/spruce/src/pages/container/EventsTable/index.tsx [hunk 1]
```diff
@@ -79,7 +79,6 @@ const EventsTable: React.FC<{}> = () => {
     defaultColumn: {
       enableColumnFilter: false,
     },
-    manualPagination: true,
```
Comment (minnakt) [inline apps/spruce/src/pages/container/EventsTable/index.tsx:82]: Was this change intentional? It looks like the page is unrelated to the packages table

File: apps/spruce/src/pages/image/PackagesTable/index.tsx [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  ImagePackagesQuery,
+  ImagePackagesQueryVariables,
+} from "gql/generated/types";
+import { IMAGE_PACKAGES } from "gql/queries";
+import { useFirstImage } from "hooks";
+
+export const PackagesTable: React.FC = () => {
+  const { [slugs.imageId]: imageId } = useParams();
+  const { image: firstImage } = useFirstImage();
+  const selectedImage = imageId ?? firstImage;
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/index.tsx:27]: let's pass down imageId as props so you don't have to call `useFirstImage`

File: apps/spruce/src/pages/image/PackagesTable/index.tsx [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  const table = useLeafyGreenTable<Package>({
+    columns,
+    data: packageEntries,
+    containerRef: tableContainerRef,
+    defaultColumn: {
+      enableColumnFilter: false,
+    },
+    manualPagination: true,
+    manualFiltering: true,
+    rowCount: totalNumPackages,
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/index.tsx:98]: just curious, does this `rowCount` do anything? It seems like it's handled by `Pagination`

File: apps/spruce/src/pages/image/PackagesTable/index.tsx [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+      columnFilters,
+    },
+    onColumnFiltersChange: setColumnFilters,
+    onPaginationChange: setPagination,
+  });
+
+  return (
+    <>
+      <SettingsCardTitle>Packages</SettingsCardTitle>
+      <SettingsCard data-cy="packages-table-card">
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/index.tsx:110]: let's remove `SettingsCardTitle` and `SettingsCard` from this component; I think this component should just contain the table

File: apps/spruce/src/pages/image/PackagesTable/index.tsx [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        <BaseTable
+          data-cy-row="packages-table-row"
+          table={table}
+          shouldAlternateRowColor
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/index.tsx:114]: can we adding `loading` and `loadingRows`?

File: apps/spruce/src/pages/image/PackagesTable/index.tsx [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            }}
+            numTotalItems={totalNumPackages}
+            currentPage={table.getState().pagination.pageIndex + 1}
+            onCurrentPageOptionChange={(value: string) => {
+              table.setPageIndex(Number(value) - 1);
+            }}
+            onBackArrowClick={() => table.previousPage()}
+            onForwardArrowClick={() => table.nextPage()}
+          />
+        </PaginationWrapper>
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/index.tsx:130]: it would be nice if we added this component to the `BaseTable` component so it could be reused for future tables (and then we could remove this code from this component)

File: apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+} from "test_utils";
+import { ApolloMock } from "types/gql";
+import { PackagesTable } from ".";
+
+describe("packages table", () => {
+  it("shows name field data", async () => {
+    const { Component } = RenderFakeToastContext(
+      <MockedProvider mocks={[imagePackagesPageOneMock]}>
+        <PackagesTable />
+      </MockedProvider>,
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx:23]: instead of redeclaring this in every test, can we have a `wrapper` component that gets passed to the `render` function? Similar to what you did in the tests for the Event Log

File: apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+    );
+    render(<Component />);
+    await waitFor(() => {
+      expect(screen.getByDataCy("packages-table-card")).toBeInTheDocument();
+    });
+    const card = screen.getByDataCy("packages-table-card");
+    const expectedNames = [
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx:80]:
```suggestion
    const expectedVersions = [
```

**Review by minnakt — CHANGES_REQUESTED** [review]
great tests!

File: apps/spruce/src/pages/image/PackagesTable/index.tsx (hunk 3 above)
Comment (sophiabarness) [reply apps/spruce/src/pages/image/PackagesTable/index.tsx:98]: The docs seem to suggest passing it in if we know the rowCount because it makes calculating the total number of entries easier for tanstack.
[Docs](https://tanstack.com/table/latest/docs/guide/pagination#pagination-row-model)

File: apps/spruce/src/pages/image/PackagesTable/index.tsx (hunk 6 above)
Comment (sophiabarness) [reply apps/spruce/src/pages/image/PackagesTable/index.tsx:130]: This seems like a good idea. I'm just a bit worried about how it would impact the BaseTables already in the code. For instance, the User Subscriptions table does pagination on top of the table, while this component does pagination below the table. Would there be a simple way of defining the order of the table and pagination?

Comment (minnakt) [reply apps/spruce/src/pages/image/PackagesTable/index.tsx]: You can use pass a prop that defines whether to use the pagination comment e.g. `usePagination` and it can be set to `false` by default to not impact any existing tables

File: apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+    }
+  });
+
+  it("shows version field data", async () => {
+    const { Component } = RenderFakeToastContext(
+      <MockedProvider mocks={[imagePackagesPageOneMock]}>
+        <PackagesTable imageId="ubuntu2204" />
+      </MockedProvider>,
+    );
+    render(<Component />);
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx:83]: You can reuse the wrapper defined above

File: apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+  it("supports name filter", async () => {
+    const user = userEvent.setup();
+    const { Component } = RenderFakeToastContext(
+      <MockedProvider
+        mocks={[imagePackagesPageOneMock, imagePackagesNameFilterMock]}
+      >
+        <PackagesTable imageId="ubuntu2204" />
+      </MockedProvider>,
+    );
+    render(<Component />);
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx:116]: same comment

File: apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+  it("supports pagination", async () => {
+    const user = userEvent.setup();
+    const { Component } = RenderFakeToastContext(
+      <MockedProvider
+        mocks={[imagePackagesPageOneMock, imagePackagesPageTwoMock]}
+      >
+        <PackagesTable imageId="ubuntu2204" />
+      </MockedProvider>,
+    );
+    render(<Component />);
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx:139]: same comment

File: apps/spruce/src/pages/image/PackagesTable/index.tsx [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+  ImagePackagesQuery,
+  ImagePackagesQueryVariables,
+} from "gql/generated/types";
+import { IMAGE_PACKAGES } from "gql/queries";
+
+type PackagesTableProps = {
+  imageId: string;
+};
+
+const PAGE_SIZE = 10;
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/index.tsx:24]: There's an existing constant called `DEFAULT_PAGE_SIZE` which you can import here

File: apps/spruce/src/pages/image/PackagesTable/index.tsx [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    },
+    {
+      header: "Manager",
+      accessorKey: "manager",
+    },
+    {
+      header: "Version",
+      accessorKey: "version",
+    },
+  ];
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/index.tsx:85]: We should move the columns outside of the component, since it doesn't depend on any of the code within the component

File: apps/spruce/src/pages/image/PackagesTable/index.tsx [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+  });
+
+  return (
+    <>
+      <BaseTable
+        data-cy-row="packages-table-row"
+        table={table}
+        shouldAlternateRowColor
+        loading={loading}
+        loadingRows={PAGE_SIZE}
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/index.tsx:113]: I think this should possibly be `pagination.pageSize`; there's no guarantee the user will always be using the default page size of 10

File: apps/spruce/src/pages/image/PackagesTable/index.tsx [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+      accessorKey: "manager",
+    },
+    {
+      header: "Version",
+      accessorKey: "version",
+    },
+  ];
+
+  const tableContainerRef = useRef<HTMLDivElement>(null);
+  const table = useLeafyGreenTable<Package>({
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/index.tsx:88]: There's a bug where if you go forward a few pages, and then search for a valid package, it doesn't reset you to the 0th page and it seems like there's no results for the search. 

There might be a setting on the table that you can add to handle this, or you can add some custom code to handle resetting the page index

File: apps/spruce/src/pages/image/PackagesTable/index.tsx [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+
+  const packageEntries = useMemo(
+    () => packagesData?.image?.packages.data ?? [],
+    [packagesData?.image?.packages.data],
+  );
+
+  const totalNumPackages = useMemo(
+    () => packagesData?.image?.packages.totalCount ?? 0,
+    [packagesData?.image?.packages.totalCount],
+  );
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/index.tsx:61]: The count doesn't appear to be correct when a filter is applied:

Screenshot 2024-08-15 at 7 36 07 PM

(Specifically there shouldn't be 103 pages, and we shouldn't be able to click the Next button)

**Review by minnakt — CHANGES_REQUESTED** [review]
(responded to your question about how to add `Pagination` to `BaseTable` in the thread above)

File: apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx [hunk 17]
```diff
@@ -0,0 +1,316 @@
...
+            {
+              manager: "pip 22.0.2 from (python 3.10)",
+              name: "bcrypt",
+              version: "3.2.0",
+            },
+          ],
+          filteredCount: 1,
+          totalCount: 15,
+        },
+      },
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx:259]: The reason why you're seeing GraphQL errors when you run the tests is because you need to include `__typename` in your mocks, like this:
```suggestion
 image: {
        __typename: "Image",
        id: "ubuntu2204",
        packages: {
           __typename: "ImagePackagesPayload",
          data: [
            {
              __typename: "Package",
              manager: "pip 22.0.2 from (python 3.10)",
              name: "bcrypt",
              version: "3.2.0",
            },
          ],
          filteredCount: 1,
          totalCount: 15,
        },
      },
```

File: apps/spruce/src/pages/image/PackagesTable/index.tsx (hunk 6 above)
Comment (sophiabarness) [reply apps/spruce/src/pages/image/PackagesTable/index.tsx]: Got it - added it along with a total count parameter to BaseTable

**Review by minnakt — CHANGES_REQUESTED** [review]

File: apps/spruce/src/components/Table/BaseTable.tsx [hunk 18]
```diff
@@ -253,3 +286,10 @@ const StyledExpandedContent = styled(ExpandedContent)`
     flex-grow: 1;
   }
 ` as typeof ExpandedContent;
+
+// @ts-expect-error
+const StyledPagination = styled(Pagination)`
+  display: flex;
+  justify-content: space-between;
```
Comment (minnakt) [inline apps/spruce/src/components/Table/BaseTable.tsx:293]: Seems like these styles are not necessary:
```suggestion
```

File: apps/spruce/src/components/Table/BaseTable.tsx [hunk 19]
```diff
@@ -80,6 +86,18 @@ export const BaseTable = forwardRef(
...
+    // Auto-redirect user to first page when filter applied.
+    useEffect(() => {
+      if (
+        table &&
+        table.getRowModel().rows.length === 0 &&
+        table.getRowCount() !== 0
+      ) {
+        table.firstPage();
+      }
+    }, [table]);
```
Comment (minnakt) [inline apps/spruce/src/components/Table/BaseTable.tsx:99]: rather than using `useEffect`, can you return to the first page exactly when the filter is applied?

File: apps/spruce/src/components/Table/BaseTable.tsx [hunk 20]
```diff
@@ -57,6 +58,9 @@ type SpruceTableProps = {
   loading?: boolean;
   /** estimated number of rows the table will have */
   loadingRows?: number;
+  usePagination?: boolean;
+  /** number of total items the table will have */
+  numTotalItems?: number;
   /** rows that will have a blue tint to represent that they are selected */
   selectedRowIndexes?: number[];
```
Comment (minnakt) [inline apps/spruce/src/components/Table/BaseTable.tsx:65]: nit. Let's keep the props in alphabetical order

File: apps/spruce/src/pages/image/PackagesTable/index.tsx [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+      pagination,
+      columnFilters,
+    },
+    onColumnFiltersChange: setColumnFilters,
+    onPaginationChange: setPagination,
+  });
+
+  if (table.getRowModel().rows.length === 0 && table.getRowCount() !== 0) {
+    table.firstPage();
+  }
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/index.tsx:81]: It looks like you forgot to delete this:
```suggestion
```

File: apps/spruce/src/pages/image/PackagesTable/index.tsx [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+
+  const packageEntries = useMemo(
+    () => packagesData?.image?.packages.data ?? [],
+    [packagesData?.image?.packages.data],
+  );
+
+  const numPackages = useMemo(
+    () => packagesData?.image?.packages.filteredCount ?? 0,
+    [packagesData?.image?.packages.filteredCount],
+  );
```
[the remark this answers is not in the document]
Comment (minnakt) [reply apps/spruce/src/pages/image/PackagesTable/index.tsx:58]: Did you apply the fall back like this?
`packagesData?.image?.packages.filteredCount ?? packagesData?.image?.packages.totalCount`

File: apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx [hunk 23]
```diff
@@ -0,0 +1,318 @@
...
+      expect(screen.queryAllByDataCy("packages-table-row")).toHaveLength(10);
+    });
+    const expectedVersions: string[] = [];
+    for (let i = 0; i < 10; i++) {
+      const packageData =
+        imagePackagesPageOneMock.result?.data?.image?.packages.data[i];
+      if (packageData && packageData.version) {
+        expectedVersions.push(packageData.version);
+      }
+    }
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx:86]: I think you could make this look a little cleaner by using the [`map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) function.

File: apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx [hunk 24]
```diff
@@ -0,0 +1,318 @@
...
+      expect(screen.queryAllByDataCy("packages-table-row")).toHaveLength(10);
+    });
+    const expectedNames: string[] = [];
+    for (let i = 0; i < 10; i++) {
+      const packageData =
+        imagePackagesPageOneMock.result?.data?.image?.packages.data[i];
+      if (packageData && packageData.name) {
+        expectedNames.push(packageData.name);
+      }
+    }
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx:46]: same comment

File: apps/spruce/src/components/Table/BaseTable.tsx [hunk 25]
```diff
@@ -1,6 +1,7 @@
-import { ForwardedRef, forwardRef } from "react";
+import { ForwardedRef, forwardRef, useEffect } from "react";
```
Comment (minnakt) [inline apps/spruce/src/components/Table/BaseTable.tsx:1]: (comment placement is random) Could we write a Cypress integration test?

**Review by minnakt — CHANGES_REQUESTED** [review]

File: apps/spruce/src/pages/image/PackagesTable/index.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (sophiabarness) [reply apps/spruce/src/pages/image/PackagesTable/index.tsx:58]: Oops got confused - this comment makes more sense now

File: apps/spruce/cypress/integration/image/build_information.ts [hunk 26]
```diff
@@ -6,4 +6,11 @@ describe("build information", () => {
       cy.contains("ubuntu1804-workstation");
     });
   });
+
+  describe("packages", () => {
+    it("should show the corresponding packages", () => {
+      cy.visit("/image/ubuntu2204");
+      cy.dataCy("packages-table-row").should("have.length", 10);
+    });
```
Comment (minnakt) [inline apps/spruce/cypress/integration/image/build_information.ts:14]: Is there anything else we can test here like applying a search, pagination? The reason the distros table test is so short is because it has no search or pagination functionalities

File: apps/spruce/src/components/Table/BaseTable.tsx [hunk 27]
```diff
@@ -57,6 +58,9 @@ type SpruceTableProps = {
   loading?: boolean;
   /** estimated number of rows the table will have */
   loadingRows?: number;
+  /** number of total items the table will have */
+  numTotalItems?: number;
+  usePagination?: boolean;
   /** rows that will have a blue tint to represent that they are selected */
   selectedRowIndexes?: number[];
```
Comment (minnakt) [inline apps/spruce/src/components/Table/BaseTable.tsx:65]: nit. to be in alphabetical order, `usePagination` should come after `selectedRowIndexes`

File: apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx [hunk 28]
```diff
@@ -0,0 +1,314 @@
...
+    );
+    render(<Component />, { wrapper });
+    await waitFor(() => {
+      expect(screen.queryAllByDataCy("packages-table-row")).toHaveLength(10);
+    });
+    const expectedNames = (
+      imagePackagesPageOneMock.result?.data?.image?.packages.data || []
+    )
+      .slice(0, 10)
+      .map((packageData) => packageData?.name);
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx:43]:
```suggestion
    const expectedNames = (
      imagePackagesPageOneMock.result?.data?.image?.packages.data || []
    ).map(({ name }) => name);
```

File: apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx [hunk 29]
```diff
@@ -0,0 +1,314 @@
...
+    );
+    render(<Component />, { wrapper });
+    await waitFor(() => {
+      expect(screen.queryAllByDataCy("packages-table-row")).toHaveLength(10);
+    });
+    const expectedVersions = (
+      imagePackagesPageOneMock.result?.data?.image?.packages.data || []
+    )
+      .slice(0, 10)
+      .map((packageData) => packageData?.version);
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/PackagesTable.test.tsx:81]:
```suggestion
    const expectedVersions = (
      imagePackagesPageOneMock.result?.data?.image?.packages.data || []
    ).map(({ version }) => version);
```

File: apps/spruce/src/pages/image/PackagesTable/index.tsx [hunk 30]
```diff
@@ -0,0 +1,135 @@
...
+      packagesData?.image?.packages.filteredCount <
+        packagesData?.image?.packages.totalCount
+    ) {
+      table.firstPage();
+    }
+  }, [
+    table,
+    packagesData?.image?.packages.filteredCount,
+    packagesData?.image?.packages.totalCount,
+  ]);
```
Comment (minnakt) [inline apps/spruce/src/pages/image/PackagesTable/index.tsx:99]: I think that the logic to reset the page should be inside of `BaseTable`, not outside of it. If the logic was kept outside, that means every time we use this component we would need to rewrite this code every time.

In `BaseTable`, the filter logic is present in the headers. Can we reset the page when a filter is applied, somewhere in the code in `BaseTable`?

**Review by minnakt — CHANGES_REQUESTED** [review]

File: apps/spruce/src/components/Table/BaseTable.tsx [hunk 31]
```diff
@@ -123,9 +129,12 @@ export const BaseTable = forwardRef(
...
                           <TableSearchPopover
                             data-cy={meta?.search?.["data-cy"]}
-                            onConfirm={(value) =>
-                              header.column.setFilterValue(value)
-                            }
+                            onConfirm={(value) => {
+                              header.column.setFilterValue(value);
+                              if (table) {
+                                table.firstPage();
+                              }
```
Comment (minnakt) [inline apps/spruce/src/components/Table/BaseTable.tsx:136]: I think it would be:
```suggestion
                              if (usePagination) {
                                table?.firstPage();
                              }
```
Could you apply this to the `TableFilterPopover` as well?

File: apps/spruce/cypress/integration/image/build_information.ts [hunk 32]
```diff
@@ -6,4 +6,41 @@ describe("build information", () => {
...
+          cy.dataCy("packages-table-row").should("have.length", 10);
+          cy.dataCy("packages-table-row")
+            .first()
+            .find("td")
+            .first()
+            .invoke("text")
+            .should((nextPackageName) => {
+              expect(nextPackageName).not.to.equal(firstPackageName);
+            });
+        });
```
Comment (minnakt) [inline apps/spruce/cypress/integration/image/build_information.ts:35]: I think it might be clearer to write it like this, including comments
```suggestion
      // First package name on first page.
      cy.dataCy("packages-table-row")
        .first()
        .find("td")
        .first()
        .invoke("text")
        .as("firstPackageName", { type: "static" });

      cy.get("[data-testid=lg-pagination-next-button]").click();

      // First package name on second page.
      cy.dataCy("packages-table-row")
        .first()
        .find("td")
        .first()
        .invoke("text")
        .as("nextPackageName", { type: "static" });

      // Package names should not be equal.
      cy.get("@firstPackageName").then((firstPackageName) => {
        cy.get("@nextPackageName").then((nextPackageName) => {
          expect(nextPackageName).not.to.equal(firstPackageName);
        });
      });
```

**Review by minnakt — APPROVED** [review]
